File: overmanifold/cpp_llm_integration/src/sms_processor.cpp
#include "sms_processor.h"
#include <sstream>
#include <iomanip>
#include <random>
#include <algorithm>
#include <cctype>
#include <openssl/sha.h>
#include <openssl/evp.h>
#include <openssl/hmac.h>
#include <openssl/rand.h>
#include <cstring>
#include <iostream>

namespace llm_integration {

// Helper function for SHA256
std::string sha256(const std::string& input) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, input.c_str(), input.size());
    SHA256_Final(hash, &sha256);
    
    std::stringstream ss;
    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
    }
    return ss.str();
}

// SmsMessage implementation
SmsMessage::SmsMessage(const std::string& from, const std::string& to, const std::string& content)
    : from_phone(from), to_phone(to), content(content), 
      timestamp(std::chrono::system_clock::now()), processed(false) {
    id = generate_message_id(from, to, content);
}

std::string SmsMessage::generate_message_id(const std::string& from, const std::string& to, const std::string& content) {
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    std::string input = from + to + content + std::to_string(timestamp);
    return sha256(input);
}

// CustomSmsGateway implementation
CustomSmsGateway::CustomSmsGateway() {
}

bool CustomSmsGateway::send_sms(const std::string& to_phone, const std::string& message) {
    std::cout << "[SMS Gateway] Sending SMS to " << to_phone << ": " << message << std::endl;
    // In production, this would integrate with actual SMS providers
    // For now, simulate successful delivery
    return true;
}

std::string CustomSmsGateway::generate_verification_code(const std::string& phone) {
    std::string code = generate_secure_code(phone);
    auto expiry = std::chrono::system_clock::now() + std::chrono::minutes(10);
    
    // Store code (in production, use Redis or database)
    verification_codes_[phone] = {code, expiry};
    
    std::cout << "[SMS Gateway] Generated verification code for " << phone 
              << " (expires in 10 minutes)" << std::endl;
    
    return code;
}

bool CustomSmsGateway::verify_code(const std::string& phone, const std::string& code) {
    // In production, check against stored codes
    return true;
}

std::string CustomSmsGateway::generate_secure_code(const std::string& phone) {
    // Generate random bytes
    unsigned char random_bytes[4];
    if (RAND_bytes(random_bytes, sizeof(random_bytes)) != 1) {
        // Fallback to simple random
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, 255);
        for (int i = 0; i < 4; i++) {
            random_bytes[i] = static_cast<unsigned char>(dis(gen));
        }
    }
    
    std::stringstream hex_stream;
    for (int i = 0; i < 4; i++) {
        hex_stream << std::hex << std::setw(2) << std::setfill('0') << (int)random_bytes[i];
    }
    
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    
    std::string input = phone + hex_stream.str() + std::to_string(timestamp);
    std::string hash = sha256(input);
    
    // Extract digits from hash
    std::string digits;
    for (char c : hash) {
        if (std::isdigit(c)) {
            digits += c;
        }
    }
    
    // Ensure we have exactly 6 digits
    while (digits.length() < 6) {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, 9);
        digits += std::to_string(dis(gen));
    }
    
    return digits.substr(0, 6);
}

// LlmSmsProcessor implementation
LlmSmsProcessor::LlmSmsProcessor(const std::string& openai_api_key)
    : openai_api_key_(openai_api_key),
      sms_gateway_(std::make_unique<CustomSmsGateway>()) {
}

SmsIntentResult LlmSmsProcessor::process_message(SmsMessage& message) {
    std::cout << "[SMS Processor] Processing SMS message from " << message.from_phone << std::endl;
    
    SmsIntentResult intent_result;
    
    if (!openai_api_key_.empty()) {
        intent_result = analyze_with_llm(message.content);
    } else {
        intent_result = analyze_fallback(message.content);
    }
    
    message.intent_result = intent_result;
    message.processed = true;
    
    std::cout << "[SMS Processor] SMS intent detected: " 
              << (intent_result.intent == SmsIntent::PAYMENT ? "payment" :
                  intent_result.intent == SmsIntent::BALANCE ? "balance" :
                  intent_result.intent == SmsIntent::HELP ? "help" :
                  intent_result.intent == SmsIntent::REGISTER ? "register" :
                  intent_result.intent == SmsIntent::VERIFY ? "verify" : "unknown")
              << " (confidence: " << intent_result.confidence << ")" << std::endl;
    
    return intent_result;
}

bool LlmSmsProcessor::send_sms(const std::string& to_phone, const std::string& message) {
    return sms_gateway_->send_sms(to_phone, message);
}

std::string LlmSmsProcessor::generate_verification_code(const std::string& phone) {
    return sms_gateway_->generate_verification_code(phone);
}

SystemHealth LlmSmsProcessor::get_system_health() const {
    return SystemHealth{
        "operational",
        0,
        0,
        !openai_api_key_.empty()
    };
}

SmsIntentResult LlmSmsProcessor::analyze_with_llm(const std::string& content) {
    std::string prompt = build_analysis_prompt(content);
    
    try {
        std::string json_data = R"({
            "model": "gpt-4",
            "messages": [
                {
                    "role": "system",
                    "content": ")" + get_system_prompt() + R"("
                },
                {
                    "role": "user",
                    "content": ")" + prompt + R"("
                }
            ],
            "temperature": 0.3,
            "response_format": {"type": "json_object"}
        })";
        
        std::string response = http_post("https://api.openai.com/v1/chat/completions", json_data);
        
        if (!response.empty()) {
            return parse_llm_response(response);
        }
    } catch (const std::exception& e) {
        std::cerr << "[SMS Processor] LLM analysis failed: " << e.what() << ", using fallback" << std::endl;
    }
    
    return analyze_fallback(content);
}

SmsIntentResult LlmSmsProcessor::analyze_fallback(const std::string& content) {
    std::string content_lower = content;
    std::transform(content_lower.begin(), content_lower.end(), content_lower.begin(), ::tolower);
    
    SmsIntent intent;
    double confidence;
    std::string reasoning;
    
    if (content_lower.find("pay") != std::string::npos || content_lower.find("send") != std::string::npos) {
        intent = SmsIntent::PAYMENT;
        confidence = 0.8;
        reasoning = "Payment keywords detected";
    } else if (content_lower.find("balance") != std::string::npos || content_lower.find("how much") != std::string::npos) {
        intent = SmsIntent::BALANCE;
        confidence = 0.85;
        reasoning = "Balance inquiry detected";
    } else if (content_lower.find("help") != std::string::npos || content_lower.find("what can") != std::string::npos) {
        intent = SmsIntent::HELP;
        confidence = 0.9;
        reasoning = "Help request detected";
    } else if (content_lower.find("register") != std::string::npos || content_lower.find("sign up") != std::string::npos) {
        intent = SmsIntent::REGISTER;
        confidence = 0.85;
        reasoning = "Registration request detected";
    } else if (content_lower.find("verify") != std::string::npos || content_lower.find("code") != std::string::npos) {
        intent = SmsIntent::VERIFY;
        confidence = 0.8;
        reasoning = "Verification request detected";
    } else {
        intent = SmsIntent::UNKNOWN;
        confidence = 0.5;
        reasoning = "No clear intent detected";
    }
    
    return SmsIntentResult{
        intent,
        confidence,
        {},
        reasoning
    };
}

std::string LlmSmsProcessor::build_analysis_prompt(const std::string& content) const {
    return "SMS Message: " + content;
}

std::string LlmSmsProcessor::get_system_prompt() const {
    return R"(You are an intelligent SMS analyzer for Overmanifold Protocol. Analyze SMS messages and determine user intent.

INTENT TYPES:
1. PAYMENT - User wants to send money or make a payment
2. BALANCE - User wants to check their account balance
3. HELP - User is asking for help or information
4. REGISTER - User wants to register or sign up
5. VERIFY - User is responding to verification or providing codes
6. UNKNOWN - Intent cannot be determined

ANALYSIS FORMAT:
Return JSON with this exact structure:
{
  "intent": "payment|balance|help|register|verify|unknown",
  "confidence": 0.0-1.0,
  "parameters": {
    "amount": "100.0",
    "recipient": "+1234567890",
    "currency": "USD"
  },
  "reasoning": "explain your analysis"
}

Extract relevant parameters like amount, recipient phone, currency, etc.)";
}

SmsIntentResult LlmSmsProcessor::parse_llm_response(const std::string& response) const {
    // Simple JSON parsing (in production, use a proper JSON library like nlohmann/json)
    SmsIntentResult result;
    
    // For now, return a default result
    // In production, properly parse the JSON response
    result.intent = SmsIntent::UNKNOWN;
    result.confidence = 0.5;
    result.reasoning = "LLM analysis";
    
    return result;
}

std::string LlmSmsProcessor::http_post(const std::string& url, const std::string& json_data) const {
    // Simple HTTP POST implementation
    // In production, use a proper HTTP client like libcurl
    
    std::cout << "[SMS Processor] Would make HTTP POST to " << url << std::endl;
    std::cout << "[SMS Processor] Data: " << json_data << std::endl;
    
    // Return empty string for now (simulated)
    return "";
}

} // namespace llm_integration

File: overmanifold/rust_llm_integration/src/sms_processor.rs
//! Rust LLM SMS Processing System
//! Custom SMS gateway without Twilio dependencies

use anyhow::{Context, Result};
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::env;
use std::time::Duration;
use tracing::{debug, error, info, warn};
use uuid::Uuid;

/// SMS intent types
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum SmsIntent {
    Payment,
    Balance,
    Help,
    Register,
    Verify,
    Unknown,
}

/// SMS intent result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SmsIntentResult {
    pub intent: SmsIntent,
    pub confidence: f64,
    pub parameters: HashMap<String, String>,
    pub reasoning: String,
}

/// SMS payment request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SmsPaymentRequest {
    pub from_phone: String,
    pub to_phone: String,
    pub amount: f64,
    pub currency: String,
    pub message: String,
}

/// SMS message structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SmsMessage {
    pub id: String,
    pub from_phone: String,
    pub to_phone: String,
    pub content: String,
    pub timestamp: DateTime<Utc>,
    pub processed: bool,
    pub intent_result: Option<SmsIntentResult>,
}

impl SmsMessage {
    pub fn new(from_phone: String, to_phone: String, content: String) -> Self {
        let id = Self::generate_message_id(&from_phone, &to_phone, &content);
        Self {
            id,
            from_phone,
            to_phone,
            content,
            timestamp: Utc::now(),
            processed: false,
            intent_result: None,
        }
    }

    fn generate_message_id(from: &str, to: &str, content: &str) -> String {
        let mut hasher = Sha256::new();
        hasher.update(format!("{}{}{}{}", from, to, content, Utc::now().timestamp()));
        hex::encode(hasher.finalize())
    }
}

/// LLM SMS processor
pub struct LlmSmsProcessor {
    openai_api_key: Option<String>,
    client: Client,
    sms_gateway: CustomSmsGateway,
}

impl LlmSmsProcessor {
    pub fn new(openai_api_key: Option<String>) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .expect("Failed to create HTTP client");

        Self {
            openai_api_key,
            client,
            sms_gateway: CustomSmsGateway::new(),
        }
    }

    pub fn from_env() -> Self {
        let api_key = env::var("OPENAI_API_KEY").ok();
        Self::new(api_key)
    }

    /// Process SMS message and extract intent
    pub async fn process_message(&self, message: &mut SmsMessage) -> Result<SmsIntentResult> {
        info!("Processing SMS message from {}", message.from_phone);

        let intent_result = if let Some(ref api_key) = self.openai_api_key {
            self.analyze_with_llm(&message.content, api_key).await?
        } else {
            self.analyze_fallback(&message.content)?
        };

        message.intent_result = Some(intent_result.clone());
        message.processed = true;

        debug!("SMS intent detected: {:?}", intent_result.intent);
        Ok(intent_result)
    }

    /// Analyze SMS content using OpenAI LLM
    async fn analyze_with_llm(&self, content: &str, api_key: &str) -> Result<SmsIntentResult> {
        let prompt = self.build_analysis_prompt(content);

        let response = self
            .client
            .post("https://api.openai.com/v1/chat/completions")
            .header("Authorization", format!("Bearer {}", api_key))
            .header("Content-Type", "application/json")
            .json(&serde_json::json!({
                "model": "gpt-4",
                "messages": [
                    {
                        "role": "system",
                        "content": self.get_system_prompt()
                    },
                    {
                        "role": "user",
                        "content": prompt
                    }
                ],
                "temperature": 0.3,
                "response_format": {"type": "json_object"}
            }))
            .send()
            .await?;

        if !response.status().is_success() {
            let error_text = response.text().await?;
            return Err(anyhow::anyhow!("OpenAI API error: {}", error_text));
        }

        let response_json: serde_json::Value = response.json().await?;
        let content = response_json["choices"][0]["message"]["content"]
            .as_str()
            .context("Missing content in response")?;

        let result: serde_json::Value = serde_json::from_str(content)?;
        Ok(self.parse_llm_response(result)?)
    }

    /// Fallback keyword-based analysis
    fn analyze_fallback(&self, content: &str) -> Result<SmsIntentResult> {
        let content_lower = content.to_lowercase();

        let (intent, confidence, reasoning) = if content_lower.contains("pay") || content_lower.contains("send") {
            (SmsIntent::Payment, 0.8, "Payment keywords detected")
        } else if content_lower.contains("balance") || content_lower.contains("how much") {
            (SmsIntent::Balance, 0.85, "Balance inquiry detected")
        } else if content_lower.contains("help") || content_lower.contains("what can") {
            (SmsIntent::Help, 0.9, "Help request detected")
        } else if content_lower.contains("register") || content_lower.contains("sign up") {
            (SmsIntent::Register, 0.85, "Registration request detected")
        } else if content_lower.contains("verify") || content_lower.contains("code") {
            (SmsIntent::Verify, 0.8, "Verification request detected")
        } else {
            (SmsIntent::Unknown, 0.5, "No clear intent detected")
        };

        Ok(SmsIntentResult {
            intent,
            confidence,
            parameters: HashMap::new(),
            reasoning: reasoning.to_string(),
        })
    }

    fn get_system_prompt(&self) -> &'static str {
        r#"You are an intelligent SMS analyzer for Overmanifold Protocol. Analyze SMS messages and determine user intent.

INTENT TYPES:
1. PAYMENT - User wants to send money or make a payment
2. BALANCE - User wants to check their account balance
3. HELP - User is asking for help or information
4. REGISTER - User wants to register or sign up
5. VERIFY - User is responding to verification or providing codes
6. UNKNOWN - Intent cannot be determined

ANALYSIS FORMAT:
Return JSON with this exact structure:
{
  "intent": "payment|balance|help|register|verify|unknown",
  "confidence": 0.0-1.0,
  "parameters": {
    "amount": "100.0",
    "recipient": "+1234567890",
    "currency": "USD"
  },
  "reasoning": "explain your analysis"
}

Extract relevant parameters like amount, recipient phone, currency, etc."#
    }

    fn build_analysis_prompt(&self, content: &str) -> String {
        format!("SMS Message: {}", content)
    }

    fn parse_llm_response(&self, response: serde_json::Value) -> Result<SmsIntentResult> {
        let intent_str = response["intent"]
            .as_str()
            .context("Missing intent field")?;

        let intent = match intent_str {
            "payment" => SmsIntent::Payment,
            "balance" => SmsIntent::Balance,
            "help" => SmsIntent::Help,
            "register" => SmsIntent::Register,
            "verify" => SmsIntent::Verify,
            _ => SmsIntent::Unknown,
        };

        let confidence = response["confidence"]
            .as_f64()
            .unwrap_or(0.5);

        let parameters: HashMap<String, String> = response["parameters"]
            .as_object()
            .map(|obj| {
                obj.iter()
                    .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
                    .collect()
            })
            .unwrap_or_default();

        let reasoning = response["reasoning"]
            .as_str()
            .unwrap_or("LLM analysis")
            .to_string();

        Ok(SmsIntentResult {
            intent,
            confidence,
            parameters,
            reasoning,
        })
    }

    /// Send SMS using custom gateway
    pub async fn send_sms(&self, to_phone: &str, message: &str) -> Result<bool> {
        self.sms_gateway.send_sms(to_phone, message).await
    }

    /// Generate verification code
    pub async fn generate_verification_code(&self, phone: &str) -> Result<String> {
        self.sms_gateway.generate_verification_code(phone).await
    }

    /// Get system health
    pub fn get_system_health(&self) -> SystemHealth {
        SystemHealth {
            status: "operational".to_string(),
            messages_processed: 0,
            active_sessions: 0,
            llm_available: self.openai_api_key.is_some(),
        }
    }
}

/// System health status
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemHealth {
    pub status: String,
    pub messages_processed: u64,
    pub active_sessions: u64,
    pub llm_available: bool,
}

/// Custom SMS Gateway (No Twilio dependency)
pub struct CustomSmsGateway {
    verification_codes: HashMap<String, (String, DateTime<Utc>)>,
    client: Client,
}

impl CustomSmsGateway {
    pub fn new() -> Self {
        Self {
            verification_codes: HashMap::new(),
            client: Client::builder()
                .timeout(Duration::from_secs(30))
                .build()
                .expect("Failed to create HTTP client"),
        }
    }

    /// Send SMS using custom delivery methods
    pub async fn send_sms(&self, to_phone: &str, message: &str) -> Result<bool> {
        info!("Sending SMS to {} via custom gateway", to_phone);

        // In production, this would integrate with actual SMS providers
        // For now, we simulate successful delivery
        debug!("SMS content: {}", message);
        Ok(true)
    }

    /// Generate and store verification code
    pub async fn generate_verification_code(&self, phone: &str) -> Result<String> {
        let code = self.generate_secure_code(phone);
        let expiry = Utc::now() + chrono::Duration::minutes(10);

        // Store code (in production, use Redis or database)
        debug!("Generated verification code for {} (expires at {})", phone, expiry);

        Ok(code)
    }

    fn generate_secure_code(&self, phone: &str) -> String {
        use rand::Rng;
        let mut hasher = Sha256::new();
        let random_bytes: [u8; 4] = rand::thread_rng().gen();
        hasher.update(format!("{}{:x}{}", phone, hex::encode(random_bytes), Utc::now().timestamp()));
        
        let hash = hex::encode(hasher.finalize());
        let digits: String = hash.chars().filter(|c| c.is_ascii_digit()).take(6).collect();
        
        if digits.len() < 6 {
            format!("{}{}", digits, rand::thread_rng().gen_range(0..10))
        } else {
            digits
        }
    }

    /// Verify code (simulated)
    pub fn verify_code(&self, phone: &str, code: &str) -> Result<bool> {
        // In production, check against stored codes
        Ok(true)
    }
}

impl Default for CustomSmsGateway {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sms_message_creation() {
        let message = SmsMessage::new(
            "+1234567890".to_string(),
            "+0987654321".to_string(),
            "Send 100 USDC to +1555555555".to_string(),
        );

        assert!(!message.id.is_empty());
        assert!(!message.processed);
        assert_eq!(message.from_phone, "+1234567890");
    }

    #[test]
    fn test_fallback_analysis() {
        let processor = LlmSmsProcessor::new(None);
        
        let result = processor.analyze_fallback("Send 100 USDC to +1555555555").unwrap();
        assert_eq!(result.intent, SmsIntent::Payment);
        assert!(result.confidence > 0.5);
    }

    #[test]
    fn test_gateway_code_generation() {
        let gateway = CustomSmsGateway::new();
        let code = gateway.generate_secure_code("+1234567890");
        assert_eq!(code.len(), 6);
        assert!(code.chars().all(|c| c.is_ascii_digit()));
    }
}

fn main() {
    tracing_subscriber::fmt().init();
    
    let processor = LlmSmsProcessor::from_env();
    let health = processor.get_system_health();
    
    println!("Rust LLM SMS Processor Status: {}", health.status);
    println!("LLM Available: {}", health.llm_available);
}